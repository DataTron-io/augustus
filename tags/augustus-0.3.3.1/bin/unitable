#!/usr/bin/env python
from kernel.unitable import command
command.main()
