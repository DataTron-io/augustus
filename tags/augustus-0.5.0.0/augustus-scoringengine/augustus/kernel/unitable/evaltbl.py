#!/usr/bin/env python

# Copyright (C) 2006-2011  Open Data ("Open Data" refers to
# one or more of the following companies: Open Data Partners LLC,
# Open Data Research LLC, or Open Data Capital LLC.)
#
# This file is part of Augustus.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""EvalTable: A UniTable supporting rule based field calculation.
  
  Rules are specified as simple python assignment expressions.
  When evaluated, the local namespace of a rule is the UniTable
  itself.  Variable referenced or assigned refer to unitable fields
  unless otherwise specified as parameters to the rule.

  Fields defined by rules are not calculated until/unless referenced.
  Dependencies between rules are handled automatically on demand.

  Note that file I/O does not preserve the rules - just the basic
  UniTable data is saved, including any dynamically calculated fields.
  To force calculation of all rule based fields, use preload_rules().

  Rules may be predefined before loading the table data, or
  added dynamically as needed.

  >>> rules = Rules('c=a+b','aa=a.astype("Float")','d=aa/c')
  >>> x = EvalTable(rules)
  >>> x['a'] = [2,5,9,5]
  >>> x['b'] = [0,-1,3,5]
  >>> print x
  +-+--+
  |a|b |
  +-+--+
  |2| 0|
  |5|-1|
  |9| 3|
  |5| 5|
  +-+--+
  >>> print x['c']
  [ 2  4 12 10]
  >>> print x
  +-+--+--+
  |a|b |c |
  +-+--+--+
  |2| 0| 2|
  |5|-1| 4|
  |9| 3|12|
  |5| 5|10|
  +-+--+--+
  >>> print x['d']
  [ 1.    1.25  0.75  0.5 ]
  >>> print x
  +-+--+--+---+----+
  |a|b |c | aa| d  |
  +-+--+--+---+----+
  |2| 0| 2|2.0| 1.0|
  |5|-1| 4|5.0|1.25|
  |9| 3|12|9.0|0.75|
  |5| 5|10|5.0| 0.5|
  +-+--+--+---+----+
  >>> x.reset_rules()
  >>> print x
  +-+--+
  |a|b |
  +-+--+
  |2| 0|
  |5|-1|
  |9| 3|
  |5| 5|
  +-+--+
  >>> print x['c']
  [ 2  4 12 10]

"""


import time
from unitable import UniTable
from rules import Rule, Rules, Expr

#################################################################

class EvalTable(UniTable):
  """Evaluation Table: a UniTable where some of the columns are
  computed dynamically.
  """
  def __init__(self,rules=None,*args,**kwargs):
    UniTable.__init__(self,*args,**kwargs)
    if not isinstance(rules,Rules):
      rules = Rules(rules)
    self._rules = rules
    self._rulearg = self.as_mapping()
    self._made = {}
    self._inprog = []

  def _new_hook(self):
    return EvalTable(rules=self._rules)

  def _grow_hook(self):
    # this is a blunt way to provide the functionality for now
    self.reset_rules()
    return self

  @property
  def rules(self): return self._rules

  def add_rule(self,rule,**context):
    '''add rule(s) to current rule set'''
    context.setdefault('__',self)
    return self._rules.add_rule(rule,**context)
  add_rules = add_rule

  def rule_keys(self):
    '''return list of keys which have associated rules'''
    return self._rules.keys()

  def preload_rules(self,keys=None):
    '''precalculate values defined by rules'''
    if keys is None:
      keys = self.rule_keys()
    for key in keys:
      self[key]
    return self

  def reset_rules(self,keys=None):
    '''remove values previously calculated by rules (will be recalculated on demand)'''
    if keys is None:
      keys = self.rule_keys()
    for key in keys:
      if self._made.has_key(key):
        del self._made[key]
        try:
          del self[key]
        except KeyError:
          pass

  def _get_vector(self,key):
    # try to return existing value
    try:
      # uninitialized vectors will return None
      out = self._data[key]
    except KeyError:
      out = None
    if out is not None:
      return out
    # otherwise try to derive value
    return self._makehook(key)

  def _makehook(self,key):
    '''try to calculate rule-based value'''
    # if no rule, let KeyError raise
    rule = self._rules[key]
    assert not self._made.has_key(key), 'circular dependency for %s' % key
    self._made[key] = 0.0
    
    self._inprog.append(time.time())
    rule(self._rulearg)
    secs = time.time() - self._inprog.pop()
    self._made[key] = secs
    # adjust start time for nested interruptions
    self._inprog = [start+secs for start in self._inprog]

    return self._data[key]

  def get_rule_times(self):
    return self._made

  def str_rule_times(self,keys=None,order='time'):
    out = []
    if keys is None:
      keys = self.rule_keys()
    data = [(self._made.get(key,None),key) for key in keys]
    if order == 'time':
      data.sort(reverse=True)
    for timing,key in data:
      if timing is None:
        extra = '<not run>'
      else:
        extra = '%1.9f secs,  %9.1f ops/sec' % (timing,len(self)/timing)
      out.append('Time: %-16s %s' % (key,extra))
    return '\n'.join(out)

  def _summhook(self):
    out = []
    out.append(self.str_rule_times())
    out.append(repr(self._rules))
    return '\n'.join(out)

  def eval(self,expr,**context):
    '''evaluate an expression in the local unitable namespace'''
    context.setdefault('__',self)
    expr = Expr(expr,**context)
    return expr(self._rulearg)
