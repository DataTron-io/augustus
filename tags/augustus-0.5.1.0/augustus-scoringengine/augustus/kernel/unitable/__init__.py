#!/usr/bin/env python

# Copyright (C) 2006-2011  Open Data ("Open Data" refers to
# one or more of the following companies: Open Data Partners LLC,
# Open Data Research LLC, or Open Data Capital LLC.)
#
# This file is part of Augustus.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""
The "UniTable" is an implementation of a conceptual "Universal Table"
that is at the core of the Augustus scoring engine.  The data
structure is analogous to an R frame: a table where the columns are
vectors of equal length, but may be of different types.  It is based
on the Python numpy package and the programming interface attempts
to maintain consistency with the style established therein.

The design goal was to create a very fast, efficient object for data
shaping, model building, and scoring, both in a batch and real-time
context.  The key features are:

- A file format that matches the native machine memory storage of the
  data.  This allows for memory-mapped access to the data, eliminating
  the need for data parsing or sequential reading. 

- Fast vector operations using any number of data columns.

- Support for demand driven, rule based calculations.  Derived columns
  can be defined in terms of operations on other columns, including other
  derived columns, and will be made available when referenced.

- The ability to invoke calculations in scalar or vector mode transparently.
  Thus, one set of rule definitions can be applied to an entire
  data set in batch mode, or to individual rows incoming as real-time
  events.

- The ability to handle huge real-time data rates by automatically
  switching to vector mode when behind, and scalar mode when keeping
  up with individual input events.

########################################################################
1. unitable tool

   "unitable" converts between CSV and binary data formats. The input
   format is autodetected and may be binary or any flavor of well-formed
   CSV.  Output formats are:

     unitable --csv	# CSV output, use --sep for custom field delim
     unitable --bin	# binary output
     unitable --tbl	# pretty printed table for viewing

   Future ouput formats may include HTML, and R frames.

   Note that there may or may not be space savings by converting from
   CSV to binary format, depending on the data.  The advantage of the
   binary format is in performance gains for subsequent processing.

########################################################################
2. unitable binary file format

   The unitable binary file contains a brief text header describing
   field names and formats, followed by raw binary data.  Each field
   is stored as a contiguous vector of native machine types.  When
   "reading" a binary file, the data is not read in the traditional
   sense.  Rather, a memory mapping is established to the disk location
   and data is made available on-demand.

   The performance gains are spectacular.  For example, one huge CSV
   file that takes an hour to read, parse, and store as internal lists
   takes about 45 seconds to "read" in binary format.

########################################################################
3. unitable python data structure

   See unitable.py for instructions
   and a large list of examples.  Typical usage is:

     from augustus.unitable import UniTable
     tbl = UniTable
     tbl.fromfile(filename)

   For simple needs, tbl can be considered a dictionary where each
   column in the table is stored as a separate list of values.
   
   For maximum benefit, it is necessary to understand the python
   numpy module. Each column is actually a numpy vector and
   supports a variety of vector operations.  Also, the UniTable
   supports operations on the entire table in the style of numpy
   records.
"""

from unitable import UniTable
from evaltbl import EvalTable
from rules import Rules, Rule

