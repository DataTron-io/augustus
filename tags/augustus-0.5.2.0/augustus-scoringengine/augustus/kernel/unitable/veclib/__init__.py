#!/usr/bin/env python

# Copyright (C) 2006-2011  Open Data ("Open Data" refers to
# one or more of the following companies: Open Data Partners LLC,
# Open Data Research LLC, or Open Data Capital LLC.)
#
# This file is part of Augustus.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""
Library of pure vector functions designed to be interoperable with
numpy. Designed to provide access to all available functions in one
namespace. This should allow switching between core packages (eg. numarray
and scipy) reasonable transparently.  Also it allows for selecting
from equivalent packages according to needs (eg. different machine
arch, batch vs real-time processing, predominance of extremely
short or long vectors, and other considerations.

Typical use:

  from augustus.kernel.unitable.veclib import veclib as vec
  # vec now contains default function choices
  vec.asarray(...)   # includes numarray namespace
  vec.chgdetect(...) # and local functions

Future use (these names will likely change):

  from augustus.kernel.unitable.veclib import MakeLib, Platform
  platform = Platform(ramsize='3G')   # choose constraints and policy
  vec = MakeLib(platform)             # get custom veclib namespace

  from augustus.kernel.unitable.veclib import Tester
  tester = Tester(vec)
  tester.selftest()
  tester.list_functions()
  print tester.benchmark(funclist(...),sizes=(1,2,3,50,500))


"""

from all import registry, veclib, corelib, VecLib
from base import as_any_array, as_num_array, as_char_array, import_native

