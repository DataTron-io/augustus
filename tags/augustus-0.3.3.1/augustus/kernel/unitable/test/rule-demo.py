"""Sample python usage of EvalTable: a UniTable with rules

"""

__copyright__ = """
Copyright (C) 2005-2007  Open Data ("Open Data" refers to
one or more of the following companies: Open Data Partners LLC,
Open Data Research LLC, or Open Data Capital LLC.)

This file is part of Augustus.

Augustus is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA
"""


import sys
from itertools import izip
from augustus.kernel.unitable import EvalTable, Rules

rules = Rules(
  "is_decline = ISO_Resp_Cd=='05'",
  "is_not_decline = is_decline==0",
  """something_not_found = (
                      POS_Entry_Mode_Not_Found.astype('Bool') |
                      POS_PIN_Entry_Cpbty_Not_Found.astype('Bool') |
                      POS_Cond_Cd_Not_Found.astype('Bool') |
                      POS_Terml_Typ_Not_Found.astype('Bool') |
                      POS_Terml_Entry_Cpbty_Not_Found.astype('Bool') |
                      MOTO_ECI_Not_found.astype('Bool') |
                      POS_Env_Cd_Not_Found.astype('Bool')
                      )""",
  """something_missing = (
                      Card_Acptr_ID_Missing.astype('Bool') |
                      Card_Acptr_Nm_Missing.astype('Bool') |
                      Card_Acptr_City_Missing.astype('Bool') |
                      Card_Acptr_Ctry_Missing.astype('Bool')
                      )""",
  """something_padded = (
                      Card_Acptr_ID_Padded.astype('Bool') |
                      Card_Acptr_Nm_Padded.astype('Bool') |
                      Card_Acptr_City_Padded.astype('Bool')
                      )""",
  "something_not_avail = something_not_found | something_missing",
  "something_nonstd = something_not_avail | something_padded",
  "is_standard = something_nonstd==0",

  "is_std_decline = is_standard & is_decline",
  "is_std_ok = is_standard & is_not_decline",
  "is_nonstd_decline = something_nonstd & is_decline",
  "is_nonstd_ok = something_nonstd & is_not_decline",

  "standard = _count_ * is_standard",
  "not_standard = _count_ * something_nonstd",
  "declines = _count_ * is_decline",
  "not_declines = _count_ * is_not_decline",
  "std_declines = _count_ * is_std_decline",
  "nonstd_declines = _count_ * is_nonstd_decline",
)

def demo(filenames):
  # read each file in turn
  for filename in filenames:
    tbl = EvalTable(rules)
    tbl.fromfile(filename)

    records = tbl['_count_'].sum()

    calc_sum = ('standard','not_standard','declines','not_declines','std_declines','nonstd_declines')
    cnt = {}
    for name in calc_sum:
      cnt[name] = tbl[name].sum()
    #sanity check
    assert cnt['declines'] + cnt['not_declines'] == records, \
      "%s + %s != %s" % (cnt['declines'],cnt['not_declines'],records)
    assert cnt['standard'] + cnt['not_standard'] == records, \
      "%s + %s != %s" % (cnt['standard'],cnt['not_standard'],records)

    out = []
    rpt_names = ('standard','declines','std_declines','nonstd_declines')
    for name in rpt_names:
      out.append('%s=%6.3f%%' % (name,100.0*cnt[name]/float(records)))

    #rej_names = ('std_declines','nonstd_declines')
    #for name in rej_names:
    #  out.append('ratio_%s=%6.3f%%' % (name,100.0*cnt[name]/float(cnt['declines'])))

    print '%14s (%8d records): %s' % (filename,records,' '.join(out))


demo(sys.argv[1:])

#################################################################
# vim:sw=2:sts=2:expandtab:shiftround
