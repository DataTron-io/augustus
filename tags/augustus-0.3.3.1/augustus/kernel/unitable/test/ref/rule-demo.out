data/sample.nab (  642338 records): standard=94.838% declines= 3.592% std_declines= 3.363% nonstd_declines= 0.229%
